caiman/source_extraction/cnmf/lazy_arrays.py

```diff
@@ -0,0 +1,444 @@
+            spatial_min = self.spatial.min(axis=0)
+
+        # get the min and max of every component, and then get the min and max of the entire array
+        prods = list()
+        for t, s in iter_product([temporal_min, temporal_max], [spatial_min, spatial_max]):
+            _p = np.multiply(t, s)
+            prods.append(np.nanmin(_p))
+            prods.append(np.nanmax(_p))
+
+        self._max = np.max(prods)
+        self._min = np.min(prods)
+
+        # get the mean of each temporal component
+        temporal_mean = np.nanmean(self.temporal, axis=1)
+
+        # get the standard deviation of each temporal component
+        temporal_std = np.nanstd(self.temporal, axis=1)
+
+        # compute mean, max, min, and std projection images
+        self._mean_image = (self.spatial @ temporal_mean).reshape(frame_dims, order="F")
+        self._max_image = (self.spatial @ temporal_max).reshape(frame_dims, order="F")
+        self._min_image = (self.spatial @ temporal_min).reshape(frame_dims, order="F")
+        self._std_image = (self.spatial @ temporal_std).reshape(frame_dims, order="F")
+
+    @property
+    def spatial(self) -> np.ndarray:
+        return self._spatial
+
+    @property
+    def temporal(self) -> np.ndarray:
+        return self._temporal
+
+    @property
+    def n_components(self) -> int:
+        return self._spatial.shape[1]
+
+    @property
+    def n_frames(self) -> int:
+        return self._temporal.shape[1]
+
+    @property
+    def shape(self) -> Tuple[int, int, int]:
+        return self._shape
+
+    @property
+    def min(self) -> float:
+        return self._min
+
+    @property
+    def max(self) -> float:
+        return self._max
+
+    @property
+    def mean_image(self) -> np.ndarray:
+        """mean projection image"""
+        return self._mean_image
+
+    @property
+    def max_image(self) -> np.ndarray:
+        """max projection image"""
+        return self._max_image
+
+    @property
+    def min_image(self) -> np.ndarray:
+        """min projection image"""
+        return self._min_image
+
+    @property
+    def std_image(self) -> np.ndarray:
+        """standard deviation projection image"""
+        return self._std_image
+
+    def _compute_at_indices(self, indices: Union[int, Tuple[int, int]]) -> np.ndarray:
+        rcm = (self.spatial @ self.temporal[:, indices]).reshape(
+            self.shape[1:] + (-1,), order="F"
+        ).transpose([2, 0, 1])
+
+        if rcm.shape[0] == 1:
+            return rcm[0]  # 2d single frame
+        else:
+            return rcm
+
+    def __repr__(self):
+        r = super().__repr__()
+        return f"{r}" \
+               f"n_components: {self.n_components}"
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            raise TypeError(f"cannot compute equality against types that are not {self.__class__.__name__}")
+
+        if (self.spatial == other.spatial) and (self.temporal == other.temporal):
```

Comment:
Is there a reason not to do

`return self.spatial == other.spatial and self.temporal == other.temporal`

? Not insisting on this, just curious why it's spelled out in a bulky way here.

Reply:
It's easier to read if it's explicitly spelt out, requires less thinking while reading.